src/ipc/ipc4/handler.c

```diff
@@ -296,10 +296,16 @@ static int set_pipeline_state(uint32_t id, uint32_t cmd, bool *delayed, uint32_t
 			}
 		}
 
-		/* resource is not released by triggering reset which is used by current FW */
-		ret = pipeline_reset(host->cd->pipeline, host->cd);
-		if (ret < 0)
-			ret = IPC4_INVALID_REQUEST;
+		/*
+		 * reset the pipeline components if STOP trigger is executed in the same thread.
+		 * Otherwise, the pipeline will be reset after the STOP trigger has finished
+		 * executing in the pipeline task.
+		 */
+		if (!*delayed) {
+			ret = pipeline_reset(host->cd->pipeline, host->cd);
+			if (ret < 0)
```

Comment:
ret can be PPL_STATUS_PATH_STOP (= 1) and this is not a error for IPC3 but for IPC4 it will be  IPC4_ERROR_INVALID_PARAM ( = 1)

Reply:
im only moving the code here. Are you suggesting the original code was wrong?

Reply:
yes,  I remember I fixed some of them but this may be missed. It is not related to your PR. You can fix it or I will

Reply:
like 
````c
	ret = pipeline_reset
	if (ret < 0) {
		ret = IPC4_INVALID_REQUEST;
	} else { // ret >= 0 is not a error for IPC3, like PPL_STATUS_PATH_STOP 
		ret = 0;
	}
````

Reply:
id like to do that in a follow up. It;s a different issue than the one Im trying to fix